maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -101,34 +83,100 @@ public boolean accept(File file) {
+                   "The distribution " + downloaded + " is compressed by unsupported format. "
+                      + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            }
+        } catch (ExecutionException ee) {
+            throw new IllegalStateException(
+               "Something bad happened when the file: " + downloadedPath + " was being extracted. "
+                  + "For more information see the stacktrace", ee);
+        }
+        if (!tempFile.delete()) {
+            log.warning("failed to delete temp directory: " + tempFile);
+        }
+        System.out.println(String.format("Resolver: Successfully extracted maven binaries from %s", downloaded));
+    }
 
-                } else if (downloadedPath.endsWith(".tar.bz2")) {
-                    Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+    private File checkIfItIsAlreadyExtracted(final String downloadedZipMd5hash) {
+
+        File[] dirs = new File(MAVEN_TARGET_DIR).listFiles(new FileFilter() {
 
-                } else {
-                    throw new IllegalArgumentException(
-                            "The distribution " + downloaded + " is compressed by unsupported format. "
-                                    + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            @Override
+            public boolean accept(File file) {
+                return file.isDirectory() && downloadedZipMd5hash.equals(file.getName());
+            }
+        });
+        if (dirs != null && isExtractionFinished(dirs)) {
+            return dirs[0];
+        }
+        return null;
+    }
+
+    private boolean isExtractionFinished(File[] dirs) {
+        if (dirs.length > 0) {
+            File dir = dirs[0];
+            int count = 0;
+            while (isTempFilePresent(dir) && count < 10) {
+                if (count == 0) {
+                    System.out.println("Waiting for extraction to finish");
                 }
-            } catch (ExecutionException ee) {
-                throw new IllegalStateException(
-                        "Something bad happened when the file: " + downloadedPath + " was being extracted. "
-                                + "For more information see the stacktrace");
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                    log.warning("Problem occurred when the thread was sleeping:\n" + e.getMessage());
+                }
+                count++;
             }
+            System.out.println();
+            return !isTempFilePresent(dir);
+        } else {
+            return false;
+        }
+    }
 
-        return toExtract;
+    private boolean isTempFilePresent(File dir) {
+        final File[] files = dir.listFiles(new FilenameFilter() {
+            @Override
+            public boolean accept(File file, String fileName) {
+                return fileName.startsWith("extractionIsProcessing");
```

Comment:
I guess, that you don't need to filter the directory - you should know the whole expected path to the file.

Reply:
the same can be applied also to this https://github.com/shrinkwrap/resolver/pull/130/files#diff-ac112072885c74c94a2b8cda38cfa10eR141